### ISSUE TITLE
1.5 Release Tracker

### ISSUE THREAD

**Issue by gnb** [novaprova/novaprova#38]
Opening this ticket to document progress on a 1.5 release.

Summary
----

- A release candidate branch `1.5rc1` has been created and has CI resources attached to it.
- The Darwin port code has been rebased into the branch
- Support for ASLR on modern Linux toolchains has been added too, building on the Darwin work
- An implementation of the DWARF Line Number Program specification has been added, allowing for line numbers in stack reports.
- The RC branch requires C++11 support to compile.
- The numerous issues across the CI matrix caused by the above upheavals are being worked through.
- The CI matrix has been updated to remove platforms which have run out of vendor support and to add new releases.
- RHEL6 is still in the late stages of vendor support but has been removed from the CI matrix because of difficulties getting a sufficiently modern toolchain in the CI environment.
- Updated documentation and release notes [available for review](https://novaprova.readthedocs.io/en/1.5rc1/)
- MacOS Homebrew packaging

CI Matrix Status
----

Status as of commit 98b3891a3f7a04b5378f5b31fb160431f928ed7e

OS | Arch | Status
-- | ---- | ------
RHEL_7 | x86_64 | :white_check_mark:
SLE_15_SP1 |  x86_64 | :white_check_mark:
openSUSE_13.1 |  i586 | :white_check_mark:
openSUSE_13.1 | x86_64 | :white_check_mark:
openSUSE_13.2 |  i586 | :white_check_mark:
openSUSE_13.2 | x86_64 | :white_check_mark:
openSUSE_Leap_15.1 |  x86_64 | :white_check_mark:
openSUSE_Leap_15.2 | x86_64 | :white_check_mark:
xUbuntu_16.04 | i586 | :white_check_mark:
xUbuntu_16.04 | x86_64 | :white_check_mark:
xUbuntu_18.04 | x86_64 | :white_check_mark:
xUbuntu_20.04 | x86_64 | :white_check_mark:
MacOS 10.14 | x86_64 | :white_check_mark:

Remaining Work
----

- Figuring out the release procedure again and writing it down dammit
- Documenting known problems in GH issues and release notes

Q & A
----

- Q: why the long gap since the 1.3 release?
  A: `$DAY_JOB` was a black hole sucking in all my time.
- Q: Why is this release numbered 1.5? The last release I saw was 1.3, what happened to 1.4?
  A: The 1.4 release partially exists in a confusing way which is hard to discover and reconcile given the large time gap since the 1.3 release. To ensure a clean transition we will skip 1.4 entirely.
- Q: is there an ARM port?  I hear the new MacBooks are ARM based.
  A: not yet.
- Q: Where's the MacOS CI?  I don't see it on OpenSUSE Build Service.
  A: There's an experimental [Travis CI setup](https://travis-ci.org/github/novaprova/novaprova)

**Comment by cpackham** [comment]
Hi Greg,

Let me know if there's anything I can do to help the release along.

**Comment by gnb** [comment]
Hey Chris,

For this release I really just need
- testing and bug reports, as you've been doing, thanks
- somebody to proof-read the updated documentation when that exists

After this release, I could use some help with the CI infrastructure.  OBS is getting more and more frustrating, and I want to see if I can move the entirety of the CI to Travis using something like (this magnificent .travis.yml)[https://github.com/dreibh/netperfmeter/blob/master/.travis.yml]

**Comment by da4089** [comment]
That netperfmeter Travis setup is rather impressive!

What platforms do you want to test Greg?  Supported Ubuntu LTS/RHEL?  recent Debian?  recent (Open)SuSE?  a couple of recent macOS?  a couple of recent FreeBSD?  NetBSD/OpenBSD?  Alpine?  Atomic?  Arch?  Poky/Yocto?

Depending on the OS choices, you'll get a fair coverage of recent GCC/clang versions, but it might be useful to have eg. Ubuntu builds for specific toolchain versions too?

Do you want to maintain a couple of 32-bit x86 builds?

**Comment by gnb** [comment]
> That netperfmeter Travis setup is rather impressive!

Yeah, it's a work of art.  I'm particularly impressed by the clarity and regularity of how they define the platform matrix.

> What platforms do you want to test Greg?

My rule of thumb has been: all vendor-supported LTS versions of major platforms (so no Arch et al) for which support exists in the NovaProva codebase (today this means Linuxen and MacOS).

But as you can see from the matrix at the top of this ticket, I've cheated this rule a couple of times.  For example, no Debian just Ubuntu.  And I had to abandon RHEL6 before it fell out of vendor support, even though the code almost certainly works on it.   These are limitations of the OpenSUSE Build Service model, not of the platform or of NovaProva.

I'm particularly worried about MacOS.  Travis seems to have MacOS 10.14, but my machine here is on 10.15 and there are significant changes between the two which required work to support.

> Do you want to maintain a couple of 32-bit x86 builds?

Oh yes. If the platform is vendor-supported in 32b, I want to also support it and build it in 32b.

Another CI dimension that is completely unexplored today is: with/without Valgrind.  Originally NovaProva had to have Valgrind at build time, now it's optional.

**Comment by gnb** [comment]
I've updated the documentation and release notes, and would appreciate it if someone could review it.  https://novaprova.readthedocs.io/en/1.5rc1/

**Comment by cpackham** [comment]
[Why Create NovaProva?](https://novaprova.readthedocs.io/en/1.5rc1/intro.html#why-create-novaprova)
The links http://git.cyrusimap.org/ appear to be dead. CUnit now has a fork at https://gitlab.com/cunity/cunit which seems more active than the old sourceforge project (not sure if you want to link to that).

[Getting Started](https://novaprova.readthedocs.io/en/1.5rc1/getting-started.html#building-and-running-tests) should the `dsymutil` be conditional on CC = clang rather than just Darwin?

There's a bit of repetition between [Getting Started](https://novaprova.readthedocs.io/en/1.5rc1/getting-started.html#building-and-running-tests) and [Setting Up the Makefile](https://novaprova.readthedocs.io/en/1.5rc1/building.html#setting-up-the-makefile). Probably helps those who skip straight that chapter but I found myself wondering if I'd followed the wrong link. I wonder if perhaps "Getting Started" could say less.

[Looping, Deadlocked, Or Slow Tests](https://novaprova.readthedocs.io/en/1.5rc1/failures.html#looping-deadlocked-or-slow-tests) is there a way of manually adjusting the timeout? If so a link or mention of it here would be useful.

[Output Formats](https://novaprova.readthedocs.io/en/1.5rc1/output-formats.html#output-formats) I recall seeing something about TAP support. Is that a thing for 1.5?

Fixtures, Mocking - read no comments.

Parameters - cool I learned something new. Need to file that one away. I've seen some tests that could benefit from this.

[Syslog Matching](https://novaprova.readthedocs.io/en/1.5rc1/c-api-ref.html#syslog-matching) the commentary on `np_syslog_ignore()` that says "Note that this is the default behaviour" but that contradicts [Messages Emitted To syslog()](https://novaprova.readthedocs.io/en/1.5rc1/failures.html#syslog) which says that any call to `syslog()` will be considered a failure.

[Release 1.5](https://novaprova.readthedocs.io/en/1.5rc1/releasenotes.html#release-1-5) "A new can be supplied at configure time, but shouldn’t be necessary" missing a word, probably should be "A new one" or "A new suppression file".

That's about it for a first pass through.

**Comment by gnb** [comment]
Thanks @cpackham !  These are all useful feedback and I will go through them this week.  A couple of quick answers:

> The links http://git.cyrusimap.org/ appear to be dead

Updated.

> CUnit now has a fork at https://gitlab.com/cunity/cunit

Updated the chapter to mention that.

> Getting Started should the dsymutil be conditional on CC = clang rather than just Darwin?

On Darwin there's a binary called `gcc` in the default `$PATH` which is actually Apple's build of clang, so I don't think that check would work correctly.

> There's a bit of repetition between Getting Started and Setting Up the Makefile. 

Agreed.  Getting Started was once a separate document which duplicated all that information, the merge into the manual needs more work.

> Looping, Deadlocked, Or Slow Tests is there a way of manually adjusting the timeout? I

There isn' t.  Feel free to raise a ticket if that would be helpful to you.

> Output Formats I recall seeing something about TAP support. Is that a thing for 1.5?

Not in 1.5.  It's the `experimental` branch and there's a lot of funky stuff in there which might be useful but needs some loving before it can be merged in.

> Parameters - cool I learned something new. Need to file that one away. I've seen some tests that could benefit from this

:)

I will look into the other two points later.

**Comment by gnb** [comment]
After a lot of struggle I managed to get the Homebrew packaging working.  There's a bottle for Catalina only, other MacOS versions build from source :)

```
$ brew info novaprova
novaprova/novaprova/novaprova: stable 1.5rc2 (bottled)
The new generation unit test framework for C
http://www.novaprova.org/
/usr/local/Cellar/novaprova/1.5rc2 (56 files, 11.7MB) *
  Poured from bottle on 2020-06-27 at 15:48:32
From: https://github.com/novaprova/homebrew-novaprova/blob/HEAD/Formula/novaprova.rb
==> Dependencies
Required: autoconf ✔, automake ✔, binutils ✔, doxygen ✔, gettext ✔, libiberty ✔, libxml2 ✔, pkg-config ✔
```

**Comment by cpackham** [comment]
Hi Greg,

Just checking in. Anything I can do to help nudge the release along? At $dayjob we've been happily running our tests for one project with the 1.5rc1 branch for a while. I was tempted to just package it up and deploy it to more projects but if a official 1.5 tarball was imminent it would make more sense to use that.
